documentation/MSBuild-Server.md

```diff
@@ -0,0 +1,54 @@
+# MSBuild Server
+
+MSBuild Server is basically an another type of node which can accept build request from clients and utilize worker nodes in current fashion to build projects. Main purpose of the server node is to avoid expensive MSBuild process start during build from tools like .NET SDK.
+
+## Communication protocol
+
+The server node uses same IPC approach as current worker nodes - named pipes. This solution allows to reuse existing code. When process starts, pipe with deterministic name is opened and waiting for commands. Client has following worfklow:
+
+1. Try to connect to server
+   - If server is not running, start new instance
+   - If server is busy, fallback to classic build 
+2. Initiate handshake
+2. Issue build command with `EntryNodeCommand` packet
+3. Read packets from pipe
+   - When `EntryNodeConsoleWrite` packet type is recieved, write content to appropriate output stream with respected coloring
+   - When `EntryNodeResponse` packet type is recieved, build is done and client writes trace message with exit code
+
+### Pipe name convention & handshake
+
+There might be multiple server processes started with different architecture, associated user, MSBuild version and another options. To quickly identify the appropriate one, server uses convention that includes these options in the name of the pipe. Name has format `MSBuildServer-{hash}` where `{hash}` is a SHA256 hashed value identifying these options.
+
+Handshake is a procedure ensuring that client is connecting to a compatible server instance. 
+
+### Packets for client-server communication
+
+Server requires to introduce new packet types for IPC.
+
+`EntryNodeCommand` contains all of the information necessary for a server to run a build.
+
+| Property name            | Type                        | Description |
+|---|---|---|
+| CommandLine              | String                      | The MSBuild command line with arguments for build |
+| StartupDirectory         | String                      | The startup directory path |
+| BuildProcessEnvironment  | Dictionary<String, String>  | Environment variables for current build |
+| Culture                  | CultrueInfo                 | The culture value for current build |
+| UICulture                | CultrueInfo                 | The UI culture value for current build |
+
+`EntryNodeConsoleWrite` contains information for console output.
+
+| Property name            | Type          | Description |
+|---|---|---|
+| Text                     | String        | The text that is written to output stream |
+| Foreground               | ConsoleColor  | The foreground color of the text |
```

Comment:
One of the request is "output fidelity". MSBuild Server doesn't have it's own window/console host, so it must pass text to client (.NET SDK) to render it. And it should look same as if you start MSBuild in it's own console host.

Reply:
I agree with the "this is just a dumb pipe and the existing MSBuild stuff does colorization" approach, because it's the only way to get custom loggers working correctly.

> I'd expect messages to be tagged with some kind of tag that has semantic meaning (e.g. Error, Fatal, etc), and then a host to convert that meaning to a color.

This is basically what happens today, but it's the _logger_ that converts `BuildEventArgs` meaning to colors, and the logger is a pluggable entity that must run in the MSBuild server side, not the client (without further changes).

Reply:
Thanks, that makes sense now that you've made the connection to the Loggers for me.  

From what I can see, there's no explicit tie to the Color in that interface, and System.ConsoleColor is a very very small subset of the allowed colors on a terminal (for example the ANSI terminal output that Spectre.Console supports is listed [here](https://spectreconsole.net/appendix/colors). Two points related to that: 
* Does MSBuild only write entire lines in one color? If so, and that changes, how does this message protocol change to allow for printing 'segments' of a line in a certain color?
* From what I can see MSBuild currently only uses ConsoleColors, but most terminals can support more colors than that. Usually detection of this support is done by probing the terminal at execution-time to see what the terminal supports. We won't be able to do that in a client/server scenario like this, so somehow the terminal output capabilities of the user's terminal would need to be included in the BuildProcessEnvironment, right?

Reply:
The _default console logger that ships with MSBuild_ a) uses `System.Console` colors (because it predates Windows VT100 support) and b) does indeed do line-at-a-time coloring. But third-party loggers can do whatever they want including prettier highlighting and the probing you're describing; I don't know if there's a good solution for that.

Reply:
Given that, should we drop the colors from this API and just make this contract purely textual output, that may or may not already have ANSI codes embedded in it? Assuming the client capabilities are passed in, loggers should be able to inspect them to determine how they want to modify that text output.

Reply:
I was thinking about it and I agree to use just text with ANSI escape sequences. Maybe we can even remove `EntryNodeConsoleWrite` completely and redirect output/error streams of the server process. It should work if the stream contains raw data with escape codes. I believe it does, but I will check it.
As Chet mentioned, we should provide details about console capabilities to the server. Instead of environment dictionary, I would prefer extra property as this is well known and defined value.

Reply:
Yeah, we just need to trip into this mode:

https://github.com/dotnet/msbuild/blob/40f49cd8d06ec3ed6933fa66183e8d1615cb2edc/src/Build/Logging/ConsoleLogger.cs#L138-L142

Then the console logger will emit ANSI codes instead of calling the Win32 console APIs and everything should Just Work As Normal.

Reply:
It's true that simple redirection in C# with standard API allows to redirect when you start new process from your code. Win32 Console API should provide functions to read from stream of already running process, but it looks ugly and it's not cross platform, so we can use pipe for sake of simplicity.